### DISCUSSION TITLE
Layout Extensions

### DISCUSSION THREAD

**Discussion by abertay-university** [directus/directus#7195] [Archive]
I've been developing a few extensions and have hit a bit of a wall with layouts. In short, how can I access the items object for my layout?

I can see an api call has been made in Chrome's dev tools that fetches the first 25 rows using the selected columns from the layoutOptions. But I can't find it in the layoutState.

I'm able to use the layoutState for the search, filters and options, and inject('system') to retrieve the Stores and access the API, but neither contain the items object.

The screenshot below shows the /items/pages?limit=25&fields.... request from directus. 
![image](https://user-images.githubusercontent.com/31690037/128242299-33ff54c7-4dea-4551-9cba-d2a7e1afe2c3.png)

To get around this, I've copied the code from the consumable use-items which allows me to use the function useItems (with prefix site_) to get the items object and also the getItems function using the inject('system') API.

You'll see in the same screenshot my api call /items/pages?limit=-1&fields.... immediately after the native one.

This is my code in index.js
```js
const { sort, limit, page, fields, fieldsWithRelational } = site_useItemOptions();
const {
  items,
  loading,
  error,
  totalPages,
  itemCount,
  totalCount,
  site_getItems,
} = site_useItems(collection, {
  sort,
  limit,
  page,
  fields: fieldsWithRelational,
  filters: filters,
  searchQuery: searchQuery,
});
```

Following this code above is a copy of the consumables with the API and Store directed to the injected system.

This works for me however it's not ideal to have duplicate api requests especially once I scale this up.

This brings me back around to my question, how can I access the items object natively for my layout?

Also, can I make changes the api query before it's sent? Such as the limit and/or filters?

Lastly, I wanted to say how amazing this system is. I'm blown away by the dedication and quality of your team. Directus 9 will be a ground-breaking release. I hope my little contributions have helped.

Thanks
Tim

**Comment by rijkvanzanten** [comment]
@nickrum Any insights?

> Lastly, I wanted to say how amazing this system is. I'm blown away by the dedication and quality of your team. Directus 9 will be a ground-breaking release. I hope my little contributions have helped.

Thanks! ❤️

**Comment by nickrum** [comment]
The items object is not exposed to layouts on purpose to allow for maximum flexibility inside layouts. Those requests you are seeing in Chrome's dev tools are likely caused by every layout type being initialized in parallel and not just the currently selected one. This is a limitation of our current layout system.

We will expose those composables that are useful for building layouts in the future to make working with custom layouts much more enjoyable.

> Also, can I make changes the api query before it's sent? Such as the limit and/or filters?

Couldn't you just alter the limit/filters options passed to `site_useItems`?

**Reply by abertay-university** [reply]
Thanks @nickrum for your response. That explains why I couldn't find it.

Since I created my own composable `site_useItems` (copy of useItems) I changed the default limit to -1 and added a default filter as well.

My main concern is around performance. Whenever I load my custom layout, I can see the native api call followed by my own api call. Both fetch the same fields and get the same response except for the limit which I've changed for mine. I also have my custom layout available as a m2o field. I see the repeated api call there as well.

Ideally, I'd like to utilise the existing api call and use the LayoutOption to configure any limit/filter changes.

Since the API call is already happening, maybe the response could be included in the layoutState? This way it's available if we want it.

**Reply by nickrum** [reply]
This is currently not possible, as the other API calls you are seeing aren't really "native". They are issued by the other layout types and thus not accessible when the layout state is initialized. Even if that would be possible, it wouldn't help you much, as it is not possible to widen limit/filter (as in _make them more general_) after the API call has been made.

We are currently investigating how to prevent those dupicate calls from happening.

**Reply by abertay-university** [reply]
Impressive work as always. Thanks @nickrum

**Reply by jagged3dge** [reply]
@abertay-university 
May I know how ARE you making the API calls in your custom layout? 
I'm hitting the same issue as the OP.

From the looks of it, it seems you're extending the Tabular layout for something. 
I'm trying to extend the Tabular layout to add a column that allows a dropdown button with custom menu actions that call some custom endpoints for each item.

**Reply by abertay-university** [reply]
Hi @jagged3dge 
From my experience it's really difficult to extend the tabular layout. It seems we are given a blank canvas to do as we please.
What you are describing sound more like a custom display https://docs.directus.io/extensions/displays/
You can add a new field called "actions" and maybe make it read-only JSON with a default value of options or just leave empty and code the options into the extension. Then set the display of that field to your custom display. Try to make use of the existing templates for dropdown menus and links to save some time coding.

**Reply by jagged3dge** [reply]
That is a very interesting approach to a solution. Thank you @abertay-university for the suggestion. 👍
